runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/streaming/harness/FanOutStreamingEngineWorkerHarness.java

```diff
@@ -281,11 +293,14 @@ private void consumeWorkerMetadata(WindmillEndpoints windmillEndpoints) {
   private synchronized void consumeWindmillWorkerEndpoints(WindmillEndpoints newWindmillEndpoints) {
     // Since this is run on a single threaded executor, multiple versions of the metadata maybe
     // queued up while a previous version of the windmillEndpoints were being consumed. Only consume
-    // the endpoints if they are the most current version.
+    // the endpoints if they are the most current version, or if the endpoint type is different
+    // from currently active endpoints.
     synchronized (metadataLock) {
-      if (newWindmillEndpoints.version() < pendingMetadataVersion) {
+      if (newWindmillEndpoints.version() < pendingMetadataVersion
+          && newWindmillEndpoints.endpointType() == activeEndpointType) {
```

Comment:
ditto, should be 

(version < pending) || (version==pending && type=activeType)